src/messages/MOSDOp.h

```diff
@@ -242,14 +236,14 @@ struct ceph_osd_request_head {
       ::encode_nohead(snaps, payload);
     } else {
       header.version = HEAD_VERSION;
-      ::encode(client_inc, payload);
+      ::encode(pgid, payload);
```

Comment:
Right. You'll notice that MOSDOp::encode_payload() already has a features parameter passed in; the Messenger provides that based on what it gets from the other end of a given connection.
So you'd need to define a new feature bit in ceph/src/include/ceph_features.h (possibly it will share the same actual value as another feature, depending on when it gets merged and what it is) and then check for that feature when encoding to see whether to encode the new (reordered) message format or the old one. :)